File: tetris.cpp
/** tetris.py - Standard simple tetris game.
 * author antifin 2012
 * version 1.0.0
 * license WTFPLv2
 */

#include <SDL2/SDL.h>
#include <chthon2/map.h>
#include <chthon2/pixmap.h>
#include <chthon2/log.h>
#include <map>
#include <set>
#include <vector>

enum class Color {
	BACK = 0,
	FORE = 0xffffff,
	SHADOW = 0x101010,
	I = 0x00ffff,
	J = 0x0000ff,
	L = 0xff8000,
	O = 0xffff00,
	S = 0x00ff00,
	T = 0xff00ff,
	Z = 0xff0000
};

enum {
	CUP_WIDTH = 10,
	CUP_HEIGHT = 22,
	SCORE_LEVEL = 10, // Rows for one speed level.
	START_SPEED = 500,
	SPEED_MODIFIER = 10
};

typedef Chthon::Map<char> Face;
typedef std::vector<Face> Array;

class Figure {
public:
	enum { W = 4, H = 4 };
	Color color;
	Array array;

	Figure(Color figure_color, const Array & figure_array);
	const Face & face() const;
	void rotate();
	Figure clone() const;
private:
	int current;
};

Figure::Figure(Color figure_color, const Array & figure_array)
	: color(figure_color), array(figure_array), current(0)
{
}

const Face & Figure::face() const
{
	return array[current];
}

void Figure::rotate()
{
	if(++current >= array.size()) {
		current = 0;
	}
}

Figure Figure::clone() const
{
	return Figure(color, array);
}


Array from_points(const std::vector<std::vector<Chthon::Point>> & figure_as_points)
{
	Array array;
	for(const std::vector<Chthon::Point> & rotational_pose : figure_as_points) {
		Face rows(Figure::W, Figure::H, false);
		for(const Chthon::Point & pos : rotational_pose) {
			rows.cell(pos) = true;
		}
		array.push_back(rows);
	}
	return array;
}

std::vector<Figure> make_figures()
{
	std::vector<Figure> result;
	result.push_back(Figure(Color::O, from_points({ {{1,1},{2,2},{1,2},{2,1}} })));
	result.push_back(Figure(Color::I, from_points({ {{1,0},{1,1},{1,2},{1,3}}, {{0,1},{1,1},{2,1},{3,1}} })));
	result.push_back(Figure(Color::J, from_points({ {{0,1},{1,1},{2,1},{2,2}}, {{1,0},{1,1},{1,2},{2,0}}, {{0,1},{1,1},{2,1},{0,0}}, {{1,0},{1,1},{1,2},{0,2}} })));
	result.push_back(Figure(Color::L, from_points({ {{0,1},{1,1},{2,1},{0,2}}, {{1,0},{1,1},{1,2},{2,2}}, {{0,1},{1,1},{2,1},{2,0}}, {{1,0},{1,1},{1,2},{0,0}} })));
	result.push_back(Figure(Color::S, from_points({ {{0,1},{1,1},{1,0},{2,0}}, {{0,0},{0,1},{1,1},{1,2}} })));
	result.push_back(Figure(Color::T, from_points({ {{0,1},{1,1},{2,1},{1,2}}, {{1,0},{1,1},{1,2},{2,1}}, {{0,1},{1,1},{2,1},{1,0}}, {{1,0},{1,1},{1,2},{0,1}} })));
	result.push_back(Figure(Color::Z, from_points({ {{0,0},{1,1},{1,0},{2,1}}, {{1,0},{0,1},{1,1},{0,2}} })));
	return result;
}

Chthon::Point shifted(const Chthon::Point & pos, const Chthon::Point & shift)
{
	return pos + shift;
}

typedef Chthon::Map<Color> Cup;

bool valid_pos(const Cup & cup, const Chthon::Point & pos, const Figure & figure)
{
	for(int col = 0; col < Figure::W; ++col) {
		for(int row = 0; row < Figure::H; ++row) {
			if(figure.face().cell(col, row)) {
				if(row + pos.y < 0 or row + pos.y >= CUP_HEIGHT) {
					return false;
				}
				if(col + pos.x < 0 or col + pos.x >= CUP_WIDTH) {
					return false;
				}
				if(cup.cell(col + pos.x, row + pos.y) != Color::BACK) {
					return false;
				}
			}
		}
	}
	return true;
}

bool may_be_fixated(const Cup & cup, const Chthon::Point & pos, const Figure & figure)
{
	return !valid_pos(cup, shifted(pos, {0, 1}), figure);
}

void fixate(const Figure & figure, const Chthon::Point & pos, Cup & cup)
{
	for(int col = 0; col < Figure::W; ++col) {
		for(int row = 0; row < Figure::H; ++row) {
			if(figure.face().cell(col, row)) {
				cup.cell(col + pos.x, row + pos.y) = figure.color;
			}
		}
	}
}

std::set<int> get_shadow(const Cup & cup, const Chthon::Point & pos, const Figure & figure)
{
	std::set<int> shadow;
	for(int col = 0; col < Figure::W; ++col) {
		for(int row = 0; row < Figure::H; ++row) {
			if(figure.face().cell(col, row)) {
				shadow.insert(pos.x + col);
				break;
			}
		}
	}
	return shadow;
}

int clear_filled_rows(Cup & cup)
{
	int lines_cleared = 0;
	Cup new_cup(cup.width(), cup.height(), Color::BACK);
	int new_cup_row = CUP_HEIGHT - 1;
	for(int row = CUP_HEIGHT - 1; row >= 0; --row) {
		bool has_gap = false;
		for(int col = 0; col < CUP_WIDTH; ++col) {
			if(cup.cell(col, row) == Color::BACK) {
				has_gap = true;
				break;
			}
		}
		if(has_gap) {
			for(int col = 0; col < CUP_WIDTH; ++col) {
				new_cup.cell(col, new_cup_row) = cup.cell(col, row);
			}
			++new_cup_row;
		} else {
			++lines_cleared;
		}
	}
	return lines_cleared;
}

int main()
{
	srand(time(nullptr));
	std::vector<Figure> figures = make_figures();

	Cup cup(CUP_WIDTH, CUP_HEIGHT);
	Chthon::Point start_pos = {(CUP_WIDTH - Figure::W) / 2, 0};
	Figure current = figures[rand() % figures.size()];
	Figure next_figure = figures[rand() % figures.size()];

	SDL_Init(SDL_INIT_EVERYTHING);
	SDL_Window * window = SDL_CreateWindow(
			"Catris",
			SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED,
			0, 0,
			SDL_WINDOW_FULLSCREEN_DESKTOP
			);
	SDL_ShowCursor(0);
	SDL_Rect screen_size;
	screen_size.x = 0;
	screen_size.y = 0;
	SDL_GetWindowSize(window, &screen_size.w, &screen_size.h);
	int cell_width = std::min(screen_size.w / CUP_WIDTH, screen_size.h / CUP_HEIGHT);

	SDL_Renderer * renderer = SDL_CreateRenderer(window, -1, 0);
	SDL_SetRenderDrawColor(renderer, 0, 0, 0, 255);
	SDL_SetRenderDrawBlendMode(renderer, SDL_BLENDMODE_BLEND);

	// make surfaces
	SDL_Surface * s = SDL_CreateRGBSurface(SDL_SWSURFACE, cell_width * CUP_WIDTH, cell_width * CUP_HEIGHT, 32, 0, 0, 0, 0); //Creating the surface.
	SDL_FillRect(s, NULL, (unsigned)Color::BACK);
	SDL_Texture * cup_surface = SDL_CreateTextureFromSurface(renderer, s);
	SDL_Rect cup_rect = {(screen_size.w - s->w) / 2, (screen_size.h - s->h) / 2, s->w, s->h};

	s = SDL_CreateRGBSurface(SDL_SWSURFACE, cell_width, cell_width * CUP_HEIGHT, 32, 0, 0, 0, 0); //Creating the surface.
	SDL_FillRect(s, NULL, (unsigned)Color::SHADOW);
	SDL_Texture * shadow = SDL_CreateTextureFromSurface(renderer, s);

	std::map<Color, SDL_Texture*> bricks;
	std::vector<Color> colors = { Color::I, Color::J, Color::L, Color::O, Color::S, Color::T, Color::Z, Color::BACK, Color::FORE, Color::SHADOW };
	for(Color color : colors) {
		s = SDL_CreateRGBSurface(SDL_SWSURFACE, cell_width, cell_width, 32, 0, 0, 0, 0); //Creating the surface.
		SDL_FillRect(s, NULL, (unsigned)Color::BACK);
		SDL_Rect r = {1, 1, s->w - 1, s->h - 1};
		SDL_FillRect(s, &r, (unsigned)color);
		bricks[color] = SDL_CreateTextureFromSurface(renderer, s);
	}
	s = SDL_CreateRGBSurface(SDL_SWSURFACE, cell_width * Figure::W, cell_width * Figure::H, 32, 0, 0, 0, 0); //Creating the surface.
	SDL_Texture * next_view = SDL_CreateTextureFromSurface(renderer, s);

	bool playing = true;
	bool quit = false;
	bool paused = false;
	int speed = 1;
	int score = 0;
	Chthon::Point figure_pos = start_pos;
	Uint32 previous_time = SDL_GetTicks();
	SDL_Event event;
	while(!quit) {
		// Drawing.
		SDL_SetRenderDrawColor(renderer, Chthon::get_red(Chthon::Color(Color::BACK)), Chthon::get_green(Chthon::Color(Color::BACK)), Chthon::get_blue(Chthon::Color(Color::BACK)), 255);
		SDL_RenderClear(renderer);
		std::set<int> shadows = get_shadow(cup, figure_pos, current);
		for(int col : shadows) {
			SDL_Rect shadow_rect = {cell_width * col, cup_rect.y, cell_width, cup_rect.h};
			shadow_rect.x += cup_rect.x;
			shadow_rect.y += cup_rect.y;
			SDL_RenderCopy(renderer, shadow, NULL, &shadow_rect);
		}
		// If playing and not paused: colors; else: draw gray content and figure.
		for(int row = 0; row < CUP_HEIGHT; ++row) {
			for(int col = 0; col < CUP_WIDTH; ++col) {
				if((unsigned)cup.cell(col, row)) {
					SDL_Rect cell_rect = {col * cell_width, row * cell_width, cell_width, cell_width};
					cell_rect.x += cup_rect.x;
					cell_rect.y += cup_rect.y;
					if(!playing) {
						SDL_RenderCopy(renderer, bricks[Color::SHADOW], NULL, &cell_rect);
					} else if(paused){
						SDL_RenderCopy(renderer, bricks[Color::FORE], NULL, &cell_rect);
					} else {
						SDL_RenderCopy(renderer, bricks[cup.cell(col, row)], NULL, &cell_rect);
					}
				}
			}
		}
		for(int row = 0; row < Figure::H; ++row) {
			for(int col = 0; col < Figure::W; ++col) {
				if(current.face().cell(col, row)) {
					SDL_Rect cell_rect = {(figure_pos.x + col) * cell_width, (figure_pos.y + row) * cell_width, cell_width, cell_width};
					cell_rect.x += cup_rect.x;
					cell_rect.y += cup_rect.y;
					if(!playing) {
						SDL_RenderCopy(renderer, bricks[Color::SHADOW], NULL, &cell_rect);
					} else if(paused){
						SDL_RenderCopy(renderer, bricks[Color::FORE], NULL, &cell_rect);
					} else {
						SDL_RenderCopy(renderer, bricks[current.color], NULL, &cell_rect);
					}
				}
			}
		}
		SDL_Rect next_view_rect = {cup_rect.x + cup_rect.w + cell_width, cup_rect.y, cell_width, cell_width};
		SDL_RenderFillRect(renderer, &next_view_rect);
		for(int row = 0; row < Figure::H; ++row) {
			for(int col = 0; col < Figure::W; ++col) {
				if(next_figure.face().cell(col, row)) {
					SDL_Rect cell_rect = {col * cell_width, row * cell_width, cell_width, cell_width};
					cell_rect.x += next_view_rect.x;
					cell_rect.y += next_view_rect.y;
					if(!playing) {
						SDL_RenderCopy(renderer, bricks[Color::SHADOW], NULL, &cell_rect);
					} else if(paused){
						SDL_RenderCopy(renderer, bricks[Color::FORE], NULL, &cell_rect);
					} else {
						SDL_RenderCopy(renderer, bricks[next_figure.color], NULL, &cell_rect);
					}
				}
			}
		}
		SDL_SetRenderDrawColor(renderer, Chthon::get_red(Chthon::Color(Color::FORE)), Chthon::get_green(Chthon::Color(Color::FORE)), Chthon::get_blue(Chthon::Color(Color::FORE)), 255);
		SDL_RenderDrawRect(renderer, &cup_rect);

		SDL_RenderPresent(renderer);

		while(SDL_PollEvent(&event)) {
			if(event.type == SDL_KEYDOWN) {
				int event_key = event.key.keysym.sym;
				if(event_key == SDLK_q) {
					quit = true;
				} else if(playing) {
					if(event_key == SDLK_UP) {
						Figure rotated = current;
						rotated.rotate();
						if(valid_pos(cup, figure_pos, rotated)) {
							current = rotated;
						}
					} else if(event_key == SDLK_DOWN) {
						while(valid_pos(cup, figure_pos, current)) {
							figure_pos = shifted(figure_pos, Chthon::Point(0, 1));
						}
						figure_pos = shifted(figure_pos, Chthon::Point(0, -1));
					} else if(event_key == SDLK_LEFT) {
						if(valid_pos(cup, shifted(figure_pos, Chthon::Point(-1, 0)), current)) {
							figure_pos = shifted(figure_pos, Chthon::Point(-1, 0));
						}
					} else if(event_key == SDLK_RIGHT) {
						if(valid_pos(cup, shifted(figure_pos, Chthon::Point(1, 0)), current)) {
							figure_pos = shifted(figure_pos, Chthon::Point(1, 0));
						}
					} else if(event_key == SDLK_p) {
						paused = !paused;
					}
				}
			} else if(event.type == SDL_QUIT) {
				quit = true;
			}
		}

		if(playing && !paused) {
			Uint32 current_ticks = SDL_GetTicks();
			if(current_ticks - previous_time > (START_SPEED - SPEED_MODIFIER * speed)) {
				previous_time = current_ticks;
				if(valid_pos(cup, shifted(figure_pos, Chthon::Point(0, 1)), current)) {
					figure_pos = shifted(figure_pos, Chthon::Point(0, 1));
				}
			}

			if(may_be_fixated(cup, figure_pos, current)) {
				fixate(current, figure_pos, cup);
				int lines_cleared = clear_filled_rows(cup);
				score += lines_cleared;
				if(score >= SCORE_LEVEL) {
					speed += 1;
					score = 0;
				}

				current = next_figure;
				figure_pos = start_pos;
				if(!valid_pos(cup, figure_pos, current)) {
					playing = false;
				}
				next_figure = figures[rand() % figures.size()];
			}
		}
	}
	return 0;
}

File: tetris.py
import pygame
import random
from pygame.locals import *

def init_sdl():
	pygame.init()
	pygame.mouse.set_visible(False)
	screen = pygame.display.set_mode((640, 480)) #, FULLSCREEN)
	background = pygame.surface.Surface(screen.get_size())
	background.fill( BACKGROUND )

class Figure:
	def __init__(self):
		self.color = (0, 0, 0)
		self.array = []
		self.

def make_figures():
	create figures (array, rotation point, color)
	Figure.I : (0, 255, 255),
	Figure.J : (0, 0, 255),
	Figure.L : (255, 128, 0),
	Figure.O : (255, 255, 0),
	Figure.S : (0, 255, 0),
	Figure.T : (255, 0, 255),
	Figure.Z : (255, 0, 0)

def main():
	screen = init_sdl(colors)
	figures = make_figures()
	current = random.choice(figures)
	next_figure = random.choice(figures)

	# TODO create 10x22 cup (list of rows)
	start game
		if controls:
			q: quit
			p: pause game
			if not paused and not stopped:
				left/right: move figure, if can (overlapping, cup borders)
				down: move on the lowest possible row
				up: rotate figure using its center
		if enough time passed and not paused and not stopped:
			move current figure 1 row down, if can(overlapping, cup borders)
		if may be fixated:
			fixate figure,
			try to remove full rows and increase score, if score > some value, increase speed.
			make current = next, if overlapping, then stop game.
			make new next,

		draw cup, shadow,
		if not paused and not stopped: draw its content, draw current figure,
		else: draw gray content and figure
		draw next figure in the left-top corner (minimized).

START_SPEED = 500
SPEED_MOD = 0.9
BACKGROUND = (0, 0, 0)
SHADOW = (10, 10, 10)
FOREGROUND = (255, 255, 255)

def main():


	tetris = Tetris(SIZE)
	old_level = tetris.level()
	speed = START_SPEED
	active, playing = True, True
	while True:
		for event in pygame.event.get():
			if event.type == QUIT:
				return
			elif event.type == KEYDOWN:
				if event.key == K_q:
					return
				elif playing:
					if event.key == K_UP:
						tetris.rotate()
					elif event.key == K_DOWN:
						tetris.drop()
					elif event.key == K_LEFT:
						tetris.left()
					elif event.key == K_RIGHT:
						tetris.right()
					elif event.key == K_p:
						active = not active

		if playing and active:
			if not tetris.tick():
				playing = False

			if tetris.level() != old_level:
				speed = SPEED_MOD * speed
			old_level = tetris.level()
			
			drawing(screen, tetris, active, background)

		pygame.time.delay(speed)



class Figure:
	WIDTH = 4
	HEIGHT = 4
	SIZE = WIDTH * HEIGHT
	NONE = 0
	_ = 0
	I = 1
	J = 2
	L = 3
	O = 4
	S = 5
	T = 6
	Z = 7
	COUNT = 8

	I_MASK = [
			0, 4, 0,  0,
			1, 5, 9, 13,
			0, 6, 0,  0,
			0, 7, 0,  0
			]
	J_MASK = [
			2,  6, 10, 15,
			1,  5,  9, 15,
			0,  4,  8, 15,
			15, 15, 15, 15
			]
	O_MASK = [
			0,  1,  2,  3,
			4,  5,  6,  7,
			8,  9, 10, 11,
			12, 13, 14, 15
			]

	I_F = [
			_, I, _, _,
			_, I, _, _,
			_, I, _, _,
			_, I, _, _
			]
	J_F = [
			J, _, _, _,
			J, J, J, _,
			_, _, _, _,
			_, _, _, _
			]
	L_F = [
			_, _, L, _,
			L, L, L, _,
			_, _, _, _,
			_, _, _, _
			]
	O_F = [
			_, _, _, _,
			_, O, O, _,
			_, O, O, _,
			_, _, _, _
			]
	S_F = [
			_, S, S, _,
			S, S, _, _,
			_, _, _, _,
			_, _, _, _
			]
	T_F = [
			_, T, _, _,
			T, T, T, _,
			_, _, _, _,
			_, _, _, _
			]
	Z_F = [
			Z, Z, _, _,
			_, Z, Z, _,
			_, _, _, _,
			_, _, _, _
			]

	def __init__(self, blocks, mask):
		self.blocks = blocks[:]
		self.rotateMask = mask[:]
		self.pos = [0, 0]
	
	def is_valid(self, p):
		return (p[0] >= 0 & p[1] >= 0 & p[0] < Figure.WIDTH & p[1] < Figure.HEIGHT);

	def at(self, p):
		return self.blocks[p[0] + p[1] * Figure.WIDTH]

	def clone(self):
		new = Figure(self.blocks, self.rotateMask)
		new.pos = self.pos
		return new

	def rotated():
		new_figure = self.clone()
		new_figure.blocks = [new_figure.blocks[new_figure.rotateMask[i]] for i in range(Figure.SIZE)]
		return new_figure

FIGURES = {
		Figure.I : Figure(Figure.I_F, Figure.I_MASK),
		Figure.J : Figure(Figure.J_F, Figure.J_MASK),
		Figure.L : Figure(Figure.L_F, Figure.J_MASK),
		Figure.O : Figure(Figure.O_F, Figure.O_MASK),
		Figure.S : Figure(Figure.S_F, Figure.J_MASK),
		Figure.T : Figure(Figure.T_F, Figure.J_MASK),
		Figure.Z : Figure(Figure.Z_F, Figure.J_MASK)
		}


class Tetris:
	ROWS_FOR_LEVEL = 10

	def __init__(self, size):
		self.size = size
		self.cup = [Figure.NONE for i in range(size[0] * size[1])]
		self.figure = self._random_figure(self._start_point())
		self.next_figure = self._random_figure(self._start_point())
		self.score = 0
	
	def is_valid(self, p):
		return (p[0] >= 0 & p[1] >= 0 & p[0] < self.size[0] & p[1] < self.size[1])

	def cell(self, p):
		if not self.is_valid(p):
			return Figure.NONE
		rel = (p[0] - self.figure.pos[0], p[1] - self.figure.pos[1])
		if self.figure.is_valid(rel) and not self._at(p):
			return self.figure.at(rel)
		return self._at(p)

	def shadow(self):
		result = []
		for x in range(Figure.WIDTH):
			for y in range(Figure.HEIGHT):
				if self.figure.at([x, y]):
					result += [self.figure.pos[0] + x]
		return set(result)

	def next(self):
		return self.next_figure

	def level(self):
		return self.score / Tetris.ROWS_FOR_LEVEL;

	def tick(self):
		self.figure.pos[1] += 1
		if not self._overlapping(self.figure):
			return True
		self.figure.pos[1] -= 1

		self._add_figure(self.figure)

		self.figure = self.next_figure.clone()
		self.next_figure = self._random_figure(self._start_point())
		return not self._overlapping(self.figure)

	def drop(self):
		while not self._overlapping(self.figure):
			self.figure.pos[1] += 1
		self.figure.pos[1] -= 1

		self._add_figure(self.figure);
		self.figure = self.next_figure;
		self.next_figure = self._random_figure(self._start_point());

	def left(self):
		self.figure.pos[0] -= 1
		if self._overlapping(self.figure):
			self.figure.pos[0] += 1

	def right(self):
		self.figure.pos[0] += 1
		if self._overlapping(self.figure):
			self.figure.pos[0] -= 1

	def rotate(self):
		new_figure = self.figure.rotated()
		if not self._overlapping(new_figure):
			self.figure = new_figure

	def _at(self, p):
		return self.cup[p[0] + p[1] * self.size[0]]

	def _start_point(self):
		return [self.size[0] / 2 - Figure.WIDTH / 2, 0]

	def _random_figure(self, pos):
		f = FIGURES[random.choice(FIGURES.keys())]
		f.pos = pos
		return f

	def _overlapping(self, figure):
		for x in range(Figure.WIDTH):
			for y in range(Figure.HEIGHT):
				if self.figure.at([x, y]):
					cupPos = [self.figure.pos[0] + x, self.figure.pos[1] + y]
					if not self.is_valid(cupPos) or self._at(cupPos):
						return True
		return False;

	def _add_figure(self, figure):
		for x in range(Figure.WIDTH):
			for y in range(Figure.HEIGHT):
				cupPos = [self.figure.pos[0] + x, self.figure.pos[1] + y]
				if self.figure.at([x, y]) and self.is_valid(cupPos):
					self.cup[cupPos[0] + cupPos[1] * self.size[0]] = self.figure.at([x, y])
		self._remove_full_rows()

	def _remove_full_rows(self):
		y = self.size[1] - 1
		while y >= 0:
			blockCount = 0;
			for x in range(self.size[0]):
				if self._at([x, y]):
					blockCount += 1

			if blockCount == self.size[0]:
				for ty in reversed(range(y + 1)):
					for tx in range(self.size[0]):
						self.cup[tx + ty * self.size[0]] = self.cup[tx + (ty - 1) * self.size[0]]
				self.score += 1;
				y += 1
			y -= 1

SIZE = (10, 22)
START_SPEED = 500
SPEED_MOD = 0.9
BACKGROUND = (0, 0, 0)
SHADOW = (10, 10, 10)
FOREGROUND = (255, 255, 255)

def drawing(screen, tetris, active, background):
	cellwidth = min([screen.get_size()[0] / tetris.size[0], screen.get_size()[1] / tetris.size[1]])
	cupsize = [cellwidth * tetris.size[0], cellwidth * tetris.size[1]]
	startsize = map(lambda x, y: (x - y) / 2, screen.get_size(), cupsize)
	cellsize = (cellwidth, cellwidth)
	cuprect = pygame.Rect(startsize, cupsize)
	statuspos = [cuprect.right + cellwidth, cellwidth]
	nextfigurerect = pygame.Rect(cuprect.right + cellwidth, cellwidth * 2, Figure.WIDTH * cellwidth, Figure.HEIGHT * cellwidth)

	blocks = {
			Figure.I : (0, 255, 255),
			Figure.J : (0, 0, 255),
			Figure.L : (255, 128, 0),
			Figure.O : (255, 255, 0),
			Figure.S : (0, 255, 0),
			Figure.T : (255, 0, 255),
			Figure.Z : (255, 0, 0)
			}
	for figure in blocks:
		surface = pygame.surface.Surface(cellsize)
		surface.fill(blocks[figure])
		pygame.draw.rect(surface, BACKGROUND, surface.get_rect(), 1)
		blocks[figure] = surface

	if not active:
		darkblocks = {}
		for figure in blocks:
			darkblocks[figure] = tuple([x/4 for x in figure[blocks]])
		blocks = darkblocks

	screen.blit(background, background.get_rect())

	shadow = pygame.surface.Surface( (cellwidth, cupsize[1]) ) # cuprect.topleft
	shadow.fill(SHADOW)
	for col in tetris.shadow():
		screen.blit(shadow, (cuprect.left + cellwidth * col, cuprect.top, cellwidth, cupsize[1]))

	pygame.draw.rect(screen, FOREGROUND, cuprect, 1)

	for x in range(tetris.size[0]):
		for y in range(tetris.size[1]):
			block = tetris.cell([x, y])
			if block:
				screen.blit(blocks[block], (cuprect.left + x * cellwidth, cuprect.top + y * cellwidth, cellwidth, cellwidth))

	"""
	painter.setPen(Settings::FOREGROUND);
	painter.setBrush(Settings::BACKGROUND);
	painter.drawText(cupRect.right() + cell.width(), cell.height(),
			tr("Level: %1.").arg(tetris.level()));
	"""

	for x in range(Figure.WIDTH):
		for y in range(Figure.HEIGHT):
			block = tetris.next().at([x, y])
			if block:
				screen.blit(blocks[block], (nextfigurerect.left + x * cellwidth, nextfigurerect.top + y * cellwidth, cellwidth, cellwidth))

	pygame.display.flip()

"""
def main():
	pygame.init()
	pygame.mouse.set_visible(False)
	screen = pygame.display.set_mode((640, 480)) #, FULLSCREEN)
	background = pygame.surface.Surface(screen.get_size())
	background.fill( BACKGROUND )


	tetris = Tetris(SIZE)
	old_level = tetris.level()
	speed = START_SPEED
	active, playing = True, True
	while True:
		for event in pygame.event.get():
			if event.type == QUIT:
				return
			elif event.type == KEYDOWN:
				if event.key == K_q:
					return
				elif playing:
					if event.key == K_UP:
						tetris.rotate()
					elif event.key == K_DOWN:
						tetris.drop()
					elif event.key == K_LEFT:
						tetris.left()
					elif event.key == K_RIGHT:
						tetris.right()
					elif event.key == K_p:
						active = not active

		if playing and active:
			if not tetris.tick():
				playing = False

			if tetris.level() != old_level:
				speed = SPEED_MOD * speed
			old_level = tetris.level()
			
			drawing(screen, tetris, active, background)

		pygame.time.delay(speed)
		"""

if __name__ == "__main__":
	main()

